pages/posts/the-last-breaking-change.md

```diff
@@ -0,0 +1,65 @@
+---
+title: "The Last Breaking Change"
+date: 2023-01-27
+tags:
+  - Specification
+type: Engineering
+cover: /img/posts/2022/the-last-breaking-change/cover.webp
+authors:
+  - name: Greg Dennis
+    photo: /img/avatars/gregsdennis.webp
+    link: https://twitter.com/gregsdennis
+    byline: JSON Tooling Implementer, Specification & Tooling Architect @Postman
+excerpt: We need to make a breaking change so that we can promise we won't again.
+---
+
+As we continue to [move toward a stable specification](https://json-schema.org/blog/posts/future-of-json-schema), we have been analyzing the various components, behaviors, and features of JSON Schema to determine what can be included, what may need modification, and what (if anything) needs to be removed or replaced.  The primary question that we had was whether there were any planned changes that would cause problems for users with existing schemas, i.e. breaking changes.
+
+To answer that, we needed user input.  So we posted across the internet in order to encourage users to participate in a [particular conversation on GitHub](https://github.com/orgs/json-schema-org/discussions/295) around how we treat any supposed breaking changes for the upcoming release.  The response was very much, "If you can help it, please don't break things."  But many people also said they would be less bothered if there was a defined migration path and tooling to help.
+
+We also performed an internal survey of feature stability, involving all of the JSON Schema core team members.  We're pleased to report that the vast majority of Draft 2020-12 can be kept completely as-is.  There are a few keywords and behaviors that could use some tweaking, but mostly in ways that we feel are generally compatible with the current release.
+
+Another common sentiment among both users and team members was that JSON Schema already has a history (and reputation) of including breaking changes when releasing new versions of the specification, meaning that a schema that validates properly with one version may not validate properly with the next.  This perception is one that we need to address.  To do that, **we need to orient ourselves and the specification so that we can promise schemas written to one version will validate consistently with every subsequent version**.
```

Comment:
> This perception is one that we need to address.

To me, this wording makes it sound like it's not true that we've made breaking changes. We're trying to change the way we work, not just the way people perceive that we work. I hope that made sense.

Reply:
I'll see if I can make it clearer that the perception is accurate.

Reply:
Yes, the previous sentence does make it sound like it's a real thing, which made that sentence stand out even more for me. It sounds like it contradicts the previous sentence.

To me, the only reason to change people's perception is if their perception is different than what's actually happening. We need to change what we're doing, then if people don't recognize that things are different, then we need to address people's perception of what we're doing. I think the word just has different connotations for me than it does for you.

Reply:
How about

> This history, and the perception that comes with it, is one that we need to address.

Reply:
> To me, the only reason to change people's perception is if their perception is different than what's actually happening.

People have a negative perception of this project because we keep breaking things.  It's a correct perception, and we _do_ need to change it, but if we try to just change how people _perceive_ this project, we've failed.  That's not what I'm advocating.

I'm saying (agreeing with you) that we need to change course.  By directly addressing the cause of the perception, we are indirectly addressing the perception itself.